pkg/gce-pd-csi-driver/disk_selection.go

```diff
@@ -0,0 +1,133 @@
+package gceGCEDriver
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/pkg/common"
+	gce "sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/pkg/gce-cloud-provider/compute"
+	"sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/pkg/parameters"
+)
+
+func SelectDisk(ctx context.Context, req *csi.CreateVolumeRequest, gce gce.GCECompute) (string, error) {
+	// This should never happen in practice.
+	if req == nil {
+		return "", fmt.Errorf("CreateVolumeRequest is nil")
+	}
+
+	dts, err := getDynamicDiskTypes(req.GetParameters())
+	if err != nil {
+		return "", fmt.Errorf("failed to get disk types from request parameters: %v", err)
+	}
+
+	if isSourceVolumeSpecified(req.GetVolumeContentSource()) {
+		return getTypeFromSourceVolume(ctx, req.GetVolumeContentSource(), gce, dts)
+	}
+
+	return selectDiskTypeFromTopologies(req.GetAccessibilityRequirements(), dts), nil
+}
+
+type dynamicDiskTypes struct {
+	PD      string
+	HD      string
+	Default string
+}
+
+// Extract disk types from the CSI CreateVolumeRequest.
+func getDynamicDiskTypes(reqParams map[string]string) (*dynamicDiskTypes, error) {
+	if reqParams == nil {
+		return nil, fmt.Errorf("request parameters are nil")
+	}
+
+	pdType := strings.ToLower(reqParams[parameters.ParameterPDType])
+	if pdType == "" {
+		return nil, fmt.Errorf("missing required parameter %q", parameters.ParameterPDType)
```

Comment:
I thought we were going to default pd and hd types to pd-balanced and HdB, respectively?